internal/pkg/certificates/certificates.go

```diff
@@ -0,0 +1,144 @@
+package certificates
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/kubewarden/kubewarden-controller/internal/pkg/constants"
+	controllerCrypto "github.com/kubewarden/kubewarden-controller/internal/pkg/crypto"
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	logger = ctrl.Log.WithName("certificates")
+)
+
+// This function will create and store in secrets, when necessary, the root CA and the certificate used in the
+// controller. If the certificate is create, the process exit to force kubernetes to redeploy the pod with the
+// latest version of the certificates.
+func RenewRootCAIfNedded(ctx context.Context, mgr ctrl.Manager, deploymentsNamespace string, controllerWebhookServiceName string) error {
+	logger.Info("Start to verify certificates")
+	k8s := mgr.GetClient()
+	rootCASecret := &corev1.Secret{}
+	err := k8s.Get(ctx, client.ObjectKey{Namespace: deploymentsNamespace, Name: constants.RootCASecretName}, rootCASecret)
+	if apierrors.IsNotFound(err) {
+		logger.Info("unable to find root CA secret.")
+		CA, err := controllerCrypto.GenerateCA()
+		if err != nil {
+			logger.Error(err, "unable to create root CA.")
+			return err
+		}
+		rootCASecret.Type = corev1.SecretTypeTLS
+		certificatePEMEncoded, err := CA.PEMEncodeCertificate()
+		if err != nil {
+			logger.Error(err, "unable to PEM encode CA certificate .")
+			return err
+		}
+		privateKeyPEMEncoded, err := CA.PEMEncodePrivateKey()
+		if err != nil {
+			logger.Error(err, "unable to PEM encode CA private key .")
+			return err
+		}
+		rootCASecret.Data = map[string][]byte{
+			"tls.crt": certificatePEMEncoded.Bytes(),
+			"tls.key": privateKeyPEMEncoded.Bytes(),
+		}
+		rootCASecret.Name = constants.RootCASecretName
+		rootCASecret.Namespace = deploymentsNamespace
+		err = k8s.Create(ctx, rootCASecret)
+		if err == nil {
+			logger.Info("root CA secret created")
+			err = renewControllerCerts(ctx, k8s, CA, []string{fmt.Sprintf("%s.%s.svc.cluster.local", controllerWebhookServiceName, deploymentsNamespace), fmt.Sprintf("%s.%s.svc", controllerWebhookServiceName, deploymentsNamespace)}, deploymentsNamespace)
+			if err != nil {
+				return err
+			}
+			logger.Info("Start to update webhook with latest CA")
+			err = updateWebhooksCaBundle(ctx, k8s, certificatePEMEncoded.Bytes())
+			if err != nil {
+				return err
+			}
+			logger.Info("Certificates and webhooks updated")
+			os.Exit(constants.ControllerReturnCodeCAInitialized)
+			return nil
+		} else {
+			logger.Error(err, "unable to create root CA secret.")
+			return err
+		}
+	}
+	if err == nil {
+		logger.Info("All certificates are valid")
```

Comment:
I think this is misleading. From what I understand this is executed only if the secret that contains the root CA has been found. No verification has been done. I think we can omit this check